Example / proof of concept to achieve a combination of head-based sampling + a basic form of tail-based sampling at a span level. (#4206)
open-telemetry/opentelemetry-dotnet: docs/trace/tail-based-sampling-span-level/README.md

# Tail Based Sampling at a span level: An Example

> This is an example / proof of concept to achieve tail-based sampling at a

Reviewer: ```suggestion
This document describes one possible way to achieve a form of tail-based sampling to include all failed spans, on top of the regular head-based sampling. This is done by leveraging extensibility mechanisms on the OTel sdk, and uses a combination of a custom sampler and an activity processor.
```

^ Just suggestion. I am not a fan of the wording "proof of concept" :)

span level using the extensibility mechanisms in OpenTelemetry.NET.

This is a way to achieve a combination of:

> - Head-based sampling (based on probabilistic sampling) and

Reviewer: why is this called probabilistic sampling? We are just respecting parent decision here, right?

Author: I was trying to contrast head-based sampling (which is a biased/probabilistic sampling) with tail-based sampling (biased/non-probabilistic sampling) - tried to elaborate a bit to clarify this.

- Tail-based sampling to get all failure spans (non-probabilistic sampling).

## How does this sampling example work?

We use a hybrid approach: we do head based sampling to get a

> probabilistic subset of all spans which includes both successful spans

Reviewer: This doc is using the words Span and SpanProcessor. I'd suggest to stick with Activity and ActivityProcessor, but the very first occurrence include Span,SPanProcessor in brackets, and then just stick with Activity terminology

and failure spans. In addition, we want to capture all failure spans.
To do this, if the parent based sampler's decision is to drop it, we return
a "Record-Only" sampling result. This ensures that the Span processor
receives that span. In the span processor, at the end of a span, we check if
it is a failure span. If so, we change the decision from "Record-Only"
to set the sampled flag so that the exporter receives the span.
In this example, each span is filtered individually without consideration to any
other spans.

This is a basic form of tail-based sampling at a span level. If a span failed,
we always sample it in addition to all head-sampled spans.

## When should you consider such an option?

This is a good option if you want to get all failure spans in addition to
head based sampling. With this, you get basic span level tail-based sampling
at a SDK level without having to install any additional components.

## Tradeoffs

Tail-sampling this way involves many tradeoffs such as:

> 1. Additional memory cost: Unlike head-based sampling where the sampling

Reviewer: ```suggestion
1. Additional performance cost: Unlike head-based sampling where the sampling
```

decision is made at span creation time, in tail sampling the decision is made

> only at the end, so there is additional memory cost.

Reviewer: ```suggestion
only at the end, so there is additional memory/processing cost.
```
